src/pip/_internal/models/link.py

```diff
@@ -170,12 +170,30 @@ def _ensure_quoted_url(url: str) -> str:
     and without double-quoting other characters.
     """
     # Split the URL into parts according to the general structure
-    # `scheme://netloc/path;parameters?query#fragment`.
-    result = urllib.parse.urlparse(url)
+    # `scheme://netloc/path?query#fragment`.
+    result = urllib.parse.urlsplit(url)
     # If the netloc is empty, then the URL refers to a local filesystem path.
     is_local_path = not result.netloc
     path = _clean_url_path(result.path, is_local_path=is_local_path)
-    return urllib.parse.urlunparse(result._replace(path=path))
+    return urllib.parse.urlunsplit(result._replace(path=path))
+
+
+def _absolute_link_url(base_url: str, file_url: str) -> str:
+    """Return an absolute file URL from a simple response.
+
+    If the file URL is already absolute, the function does nothing.
+    If the file URL is relative, it is joined with the base URL.
+    """
+    # If the file URL is already absolute, joining it with the page URL
+    # will do nothing. PyPI returns absolute URLs so we can avoid expensive
+    # urljoin() calls in the common case. (It would be technically more
+    # correct to parse the file URL and check if it has a scheme, but the
+    # slow URL parsing urljoin() does is what we're trying to avoid in the
+    # first place, so we only check for the http[s]:// prefix.)
```

Comment:
@ichard26 I was indeed confused, thanks for the clarification.

Maybe a simpler comment like this would have helped my distracted self. What do you think?

```suggestion
def _absolute_link_url(base_url: str, url: str) -> str:
    """A faster implementation of urlib.parse.urljoin with a shortcut for absolute http/https URLs."""
```

Reply:
Sounds good, please see https://github.com/pypa/pip/compare/83f9a08471ffd1a0302be9990aa257ecc7357c1b..66e5e73b77bbb0c4494911cef14a9711130d0ccd (from clicking on "force-pushed").